### PR TITLE
Refactor Directory Check

### PR DESCRIPTION
**Pull request by sreeb-msft**
Reducing redundant code and using getPathProperty for checking parent chain of given path for determining whether directory or not. 

This also updates the user agents.

### CODE REVIEW

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 1]
```diff
@@ -1305,50 +1305,24 @@ private void createDirectoryMarkerBlob(final Path path,
    */
   private void checkParentChainForFile(Path path, TracingContext tracingContext,
                                        List<Path> keysToCreateAsFolder) throws IOException {
-    if (checkPathIsDirectory(path, tracingContext)) {
+
+    FileStatus fileStatus = tryGetPathProperty(path, tracingContext, true);
+    Boolean isDirectory = fileStatus != null ? fileStatus.isDirectory() : false;
+    if (isDirectory) {
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: if its a file, we should throw 409 as per exisiting code.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 2]
```diff
@@ -1305,50 +1305,24 @@ private void createDirectoryMarkerBlob(final Path path,
...
       return;
     }
     Path current = path.getParent();
     while (current != null && !current.isRoot()) {
-      if (checkPathIsDirectory(current, tracingContext)) {
+      fileStatus = tryGetPathProperty(path, tracingContext, true);
+      isDirectory = fileStatus != null ? fileStatus.isDirectory() : false;
+      if (isDirectory) {
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:1330]: same as above for 409.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java [hunk 3]
```diff
@@ -182,4 +182,20 @@ public void testVerifyGetBlobProperty() throws Exception {
...
+
+
+  @Test
+  public void testGetPathPropertyCalledOnMkdir() throws Exception {
+    AzureBlobFileSystem fs = Mockito.spy(getFileSystem());
+    AzureBlobFileSystemStore store = Mockito.spy(fs.getAbfsStore());
+    Mockito.doReturn(store).when(fs).getAbfsStore();
+
+    fs.mkdirs(new Path("/testPath"));
+    Mockito.verify(store, Mockito.atLeast(1)).tryGetPathProperty(Mockito.any(Path.class),
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java]: lets have Mockito.times()? I think it should invoke only once.

**Review by saxenapranav — CHANGES_REQUESTED** [review]

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 4]
```diff
@@ -2164,6 +2064,124 @@ public FileStatus getFileStatus(Path path, TracingContext tracingContext, boolea
...
+      if (ex.getStatusCode() == HTTP_NOT_FOUND) {
+        LOG.debug("No explicit directory/path found: {}", path);
+        return null;
+      }
+      throw ex;
+    }
+  }
+
+  /**
+   * Method to make a call to get path property based on various configs-
```
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:2103]: We should specify in comments that this function will work only for explicit paths (files/directories).

**Review by anmolanmol1234 — APPROVED** [review]

**Review by anujmodi2021 — APPROVED** [review]
LGTM
Thanks for taking changes.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java [hunk 5]
```diff
@@ -178,8 +178,24 @@ public void testVerifyGetBlobProperty() throws Exception {
     fs.mkdirs(new Path("/src/dir"));
     Mockito.verify(testClient, Mockito.times(0)).getPathStatus(Mockito.any(String.class),
             Mockito.anyBoolean(), Mockito.any(TracingContext.class));
-    Mockito.verify(testClient, Mockito.times(1)).getBlobProperty(Mockito.any(Path.class),
+    Mockito.verify(testClient, Mockito.times(3)).getBlobProperty(Mockito.any(Path.class),
```
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java]: Hope that ingress,, mkdir swtiches tested?

**Review by saxenapranav — CHANGES_REQUESTED** [review]
Since, 409 change was missed -> tests would not be there. Can you please add. If already kindly cite the tests (if yes, why they didn't pick 409 before).

Good with the code. Great code!

**Review by saxenapranav — APPROVED** [review]
Great code. Thanks for taking suggestions.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java [hunk 6]
```diff
@@ -178,8 +178,35 @@ public void testVerifyGetBlobProperty() throws Exception {
...
             Mockito.anyBoolean(), Mockito.any(TracingContext.class));
-    Mockito.verify(testClient, Mockito.times(1)).getBlobProperty(Mockito.any(Path.class),
+    Mockito.verify(testClient, Mockito.atLeast(1)).getBlobProperty(Mockito.any(Path.class),
             Mockito.any(TracingContext.class));
 
   }
+
+  @Test
+  public void testGetPathPropertyCalledOnMkdir() throws Exception {
+    AzureBlobFileSystem fs = Mockito.spy(getFileSystem());
```
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemMkDir.java:188]: No blob prefix check added

**Pull request merged**